[Status] Add support for dark color scheme (Dark Mode)

Clashsoft opened the pull request:
## Changes

This changes the status page CSS to use different font and background colors when dark mode is enabled. This works on iOS 13, macOS Mojave and Windows 10. The website also updates automatically when the theme changes while viewing -- even if the OS setting switches depending on day/night cycle.

## Motivation

More and more websites are starting to adopt support for dark theme. It fells natural to do the same with a tool related to iOS app development, where this is encouraged as well.

## Screenshots

The live version is available at https://clashsoft.github.io/swift-evolution/.

Desktop Chrome:

<img width="990" alt="Desktop Chrome Screenshot" src="https://user-images.githubusercontent.com/4145923/92520448-e3e9a380-f21b-11ea-8f0c-898f9c4f45d6.png">

iOS:

<img width="320" alt="iOS Screenshot" src="https://user-images.githubusercontent.com/4145923/92521419-7b032b00-f21d-11ea-94df-f1750055f388.png">

theblixguy commented:
cc @krilnon

benrimmington commented:
Please update the copyright, and possibly add a `color-scheme` to index.css and/or index.html:

```diff
-// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
```

```diff
+:root {
+    color-scheme: light dark;
+}
```

```diff
         <meta charset="utf-8">
+        <meta name="color-scheme" content="light dark">
         <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no">
```

benrimmington commented:
For testing at <https://Clashsoft.github.io/swift-evolution/>, I think you can temporarily change the GitHub Pages branch to `feat/darkmode` in your [Settings](https://github.com/Clashsoft/swift-evolution/settings).

Clashsoft commented:
@benrimmington I added the meta tag and updated the copyright. `color-scheme` was marked as an unknown property in CSS. https://clashsoft.github.io/swift-evolution/ works now, thanks for the suggestion!

benrimmington commented:
The HTML `color-scheme` is good for switching to dark mode ASAP, but it might not be supported in Firefox.

* <https://caniuse.com/mdn-html_elements_meta_name_color-scheme>

* <https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meta/name#Browser_compatibility>

The CSS `color-scheme` within `:root` should be correct:

* <https://webkit.org/blog/8840/dark-mode-support-in-webkit/>

* <https://www.w3.org/TR/css-color-adjust-1/#color-scheme-prop>

I'd suggest adding both (HTML and CSS).

Clashsoft commented:
@benrimmington alright, I added the `color-scheme` to the CSS as well

krilnon commented:
This seems like a reasonable thing to add. Some of the gray background colors look a bit off to me (too bright). Did you use a particular standard to pick them?

benrimmington commented:
In light mode, the `header` and `nav` backgrounds are translucent. You could try the same in dark mode.

For completeness, the `body` background in light mode could be specified as `white`.

Does the `footer` need different colors in dark mode?

Clashsoft commented:
@krilnon I agree some colors are a little off. As a first draft, I used the color scheme of a dark theme for Bootstrap apps I am working on, but it's not based on a particular standard. I'll see if I can add some tweaks to bring it more in line with the iOS dark mode colors.

@benrimmington 

> In light mode, the header and nav backgrounds are translucent. You could try the same in dark mode.

I'll add the translucency to dark mode, but I found that it doesn't work correctly in light mode:

https://github.com/apple/swift-evolution/blob/22ea189435a7dbd464a06b4c2b7d8e6e51e25495/index.css#L116-L136

This only adds the `-webkit-backdrop-filter`, but not the regular `backdrop-filter`. The `header` class does not have this problem:

https://github.com/apple/swift-evolution/blob/22ea189435a7dbd464a06b4c2b7d8e6e51e25495/index.css#L79-L80

> For completeness, the body background in light mode could be specified as white.

Ok, I'll do that.

> Does the footer need different colors in dark mode?

The footer is already dark in light mode, so I didn't change it.

Clashsoft commented:
@krilnon @benrimmington the latest commits bring the color theme in line with iOS dark mode, and also specify a background for body in light mode.

Clashsoft commented:
@benrimmington blur is now supported in dark mode and fixed in light mode.

<img width="991" alt="Screenshot 2020-09-08 at 22 06 38" src="https://user-images.githubusercontent.com/4145923/92522795-93744500-f21f-11ea-909e-6a65d048faea.png">

benrimmington commented:
The `#search-filter` looks odd in macOS Safari. I'm guessing that adding the border color has caused it to inherit a thick inset border style (from the user agent style sheet). And should its background color be `var(--background)` rather than `var(--filter-background)`?

The `rgba()` notation might have wider support than the hexadecimal `#RRGGBBAA` notation, but I'm not sure if that matters for this page.

Otherwise, looks good to me.

Clashsoft commented:
@benrimmington thanks, I applied the suggestions.

benrimmington commented:
Sorry, I think your `--filter-background` was correct for dark mode, i.e. a lighter colored search field on a darker navigation bar (except that the `#search-icon` didn't contrast enough). But `--filter-background` wasn't defined for light mode, so it probably defaulted to transparent, whereas it should have been `white`.

The `#search-icon` inline SVG is currently using `stroke="#333333"` (in its circle and path elements). If these are both changed to `stroke="currentColor"`, it should then be possible to change the color in CSS for light and dark modes. But I think the search field looks fine as it is, so feel free to ignore me.

(@krilnon is the code owner.)

krilnon commented:
It's looking better! I pinged a colleague in design about this, so I'd like to wait a bit before merging to see what input they may have here.

Clashsoft commented:
@benrimmington here's what it looks like using `currentColor`:

<img width="241" alt="Screenshot 2020-09-08 at 23 55 09" src="https://user-images.githubusercontent.com/4145923/92531521-bc500680-f22e-11ea-8bea-8ae80fdc0cc1.png">

<img width="232" alt="Screenshot 2020-09-09 at 00 05 33" src="https://user-images.githubusercontent.com/4145923/92532242-303ede80-f230-11ea-809f-53fa58eff6e4.png">

The icons now feel quite bright. I think a better option could be to use the same color as the "Search" placeholder - but that seems to come from the user agent stylesheet. Also, not using `currentColor` probably requires picking a color that works for both light and dark mode.

benrimmington commented:
For comparison, here is the search field from macOS Catalina's System Preferences toolbar:

<img width="170" alt="search-filter-light-mode" src="https://user-images.githubusercontent.com/889638/92533313-1a2e1f80-f22a-11ea-9643-6ef5278df19a.png">

<img width="170" alt="search-filter-dark-mode" src="https://user-images.githubusercontent.com/889638/92533323-20240080-f22a-11ea-9320-ace7878925e3.png">

Clashsoft commented:
@benrimmington in Preferences it looks good because the search field background is brighter and there is less contrast. Maybe the search field should be brightened here too?

See also: the search field on the iOS App Store:

![image](https://user-images.githubusercontent.com/4145923/94320207-b37c6600-ff8c-11ea-9a2c-93a0aa7b2c26.jpeg)

benrimmington commented:
> Maybe the search field should be brightened here too?

@Clashsoft That looks good to me, but I'm not a designer, so we'll have to wait and see.

benrimmington commented:
I've changed the base branch of this pull request (see the [forums topic][]).

[forums topic]: <https://forums.swift.org/t/updating-branch-names/40412>

krilnon commented:
I heard back and the main feedback to pass on was related to contrast as well. Maybe we can run the different text/background combinations against a [contrast checker](https://webaim.org/resources/contrastchecker/), fix anything that comes up, and merge.

Clashsoft commented:
I found the following contrast problems:

`.proposal-link` with `#888888` on `#000000`, which works for normal text but not the thin labels we use.

<img width="731" alt="Screenshot 2020-10-05 at 23 47 52" src="https://user-images.githubusercontent.com/4145923/95135742-3196f480-0765-11eb-9e6d-4e968afe68a0.png">

<img width="99" alt="Screenshot 2020-10-05 at 23 47 11" src="https://user-images.githubusercontent.com/4145923/95135687-17f5ad00-0765-11eb-8f71-3039d1a82585.png">

This also happens with the `filter-button` with `#111112` as the background:

<img width="256" alt="Screenshot 2020-10-05 at 23 53 01" src="https://user-images.githubusercontent.com/4145923/95136522-8f780c00-0766-11eb-83d1-e507129e1732.png">

<img width="39" alt="Screenshot 2020-10-06 at 00 00 27" src="https://user-images.githubusercontent.com/4145923/95136731-f5fd2a00-0766-11eb-9241-7c7a7bac99aa.png">

I suppose it makes sense to brighten these a little. The same applies to the search field, which I've already changed (https://github.com/apple/swift-evolution/pull/1177#issuecomment-699181288).

There may be contrast issues with the various status colors. Let me know if I should check those too.

krilnon commented:
> There may be contrast issues with the various status colors. Let me know if I should check those too.

Since there's text on the status indicators, it's probably worth doing. Then we can brighten/darken uniformly as needed. If I get some time this week I'll take a look as well.

krilnon commented:
Taking a look at this again, I think this is worth merging with all the improvements you've made so far. Thanks @Clashsoft !
